js/base-engine.js

```diff
@@ -203,6 +204,70 @@ function d20plusEngine () {
 		});
 	};
 
+	d20plus.engine.enhanceMacros = () => {
+		const $dialog = $(`.dialog[data-macroid=${d20plus.engine._lastOpenedMacroId}]`);
+		if (!d20plus.engine._lastOpenedMacroId || !$dialog[0]) return;
+		const $macro = $dialog.find(`.macro.tokenizer`);
+		const $name = $dialog.find("input.name");
+		const $checkbox = $dialog.find(".isjs")
+			.on("change", (evt) => {
+				if ($(evt.target).prop("checked")) $macro.parent().addClass("jsdialog");
+				else $macro.parent().removeClass("jsdialog");
+			});
+		const script = d20plus.engine.decodeScript($macro.val());
+		if (script) {
+			$macro.parent().addClass("jsdialog");
+			$macro.val(script);
+			$checkbox.prop("checked", true);
+		}
+		$dialog.find(".btn.testmacro").on("click", () => {
+			if (!$checkbox.prop("checked")) return;
+			const script = $macro.val();
+			$macro.val(d20plus.engine.runScript(script));
+			setTimeout(() => $macro.val(script), 100);
+		});
+		const $buttons = $dialog.parent()
+			.find(".ui-dialog-buttonpane button:not(.active)")
+			.addClass("active");
+		$buttons.on("mouseup", (evt) => {
+			if (!$name.val()) {
+				let i = 0;
+				while (d20plus.ut.getMacroByName(`Untitled${i}`)) i++;
+				$name.val(`Untitled${i}`);
+			}
+			if (!$checkbox.prop("checked") || evt.which !== 1) return;
+			$macro.val(d20plus.engine.encodeScript($macro.val()));
+		});
+	}
+
+	d20plus.engine.decodeScript = (macro) => {
+		const parts = macro.split("...");
+		if (parts.length !== 3
+			|| parts[0] !== "bs``<``"
+			|| parts[2] !== "``>``") return;
+		const script = decodeURIComponent(atob(parts[1]));
+		return script;
+	}
+
+	d20plus.engine.encodeScript = (script) => {
+		const saved = btoa(encodeURIComponent(script));
+		return `bs\`\`<\`\`...${saved}...\`\`>\`\``;
+	}
+
+	d20plus.engine.runScript = (script) => {
```

Comment:
not a fan of this `vuln` stuff. Trying to write our own security filter here is not viable. Put a bigger/more explicit "remove exec" warning on the "trust GM scripts" config option, and use that as the boundary

not sure about the `use strict` either; if that's purely for "security," then, again, I wouldn't bother

also, if we want to allow async exec, here's a cool async version:
```js
const AsyncFunction = (async function(){}).constructor;
const fn = new AsyncFunction(macro);
try {
	// Can use `.call` to enforce a specific `this` if useful (`null`, here)
	return fn.call(null);
} catch(e) {
	// ...
}
```

Reply:
I've rewritten this part and removed that attempt at security measures.
Also added `fn.call(macro)` so that `this` refers to r20 macro object.

Reply:
About `use strict` - this is mostly to enforce vars declaration (otherwise the garbage vars will fill the window scope)